File: webtool/server/routes/project.js
"use strict";

const express = require("express");
const multer = require("multer");
const fs = require("fs");
const path = require("path");

const proj = require("../lib/project");
const pipeline = require("../lib/pipeline");

const router = express.Router();

// 브라우저는 로컬 파일의 절대경로를 노출하지 않으므로(드래그앤드롭도 동일), 경로 입력 대신
// 파일 자체를 업로드받아 서버 워크스페이스에 저장한 뒤 그 경로로 NitroPacker를 실행한다.
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 512 * 1024 * 1024 },
});

// POST /api/project/unpack  multipart/form-data: romFile, projectName
router.post("/unpack", upload.single("romFile"), async (req, res) => {
  const projectName = req.body && req.body.projectName;
  const file = req.file;
  if (!file || !projectName) {
    return res.status(400).json({ error: "romFile, projectName은 필수입니다" });
  }
  try {
    proj.assertValidName(projectName);
  } catch (ex) {
    return res.status(400).json({ error: ex.message });
  }
  if (!file.originalname.toLowerCase().endsWith(".nds")) {
    return res.status(400).json({ error: `.nds 롬 파일이 아닙니다: ${file.originalname}` });
  }

  try {
    const romPath = proj.originalRomPath(projectName);
    fs.mkdirSync(proj.projectDir(projectName), { recursive: true });
    fs.writeFileSync(romPath, file.buffer);

    const unpackDir = proj.unpackDir(projectName);
    fs.mkdirSync(unpackDir, { recursive: true });
    await proj.runNitroPacker(["unpack", "-r", romPath, "-o", unpackDir, "-n", projectName]);

    const scriptDir = proj.scriptDir(projectName);
    const mesCount = fs.existsSync(scriptDir)
      ? fs.readdirSync(scriptDir).filter((f) => f.toLowerCase().endsWith(".mes")).length
      : 0;

    const state = proj.writeState(projectName, {
      romFileName: file.originalname,
      romPath,
      createdAt: new Date().toISOString(),
    });

    res.json({ ok: true, state, mesCount });
  } catch (ex) {
    res.status(500).json({ error: ex.message });
  }
});

router.get("/list", (req, res) => {
  res.json(proj.listProjects());
});

router.get("/status", (req, res) => {
  const { name } = req.query;
  if (!name) return res.status(400).json({ error: "name은 필수입니다" });
  const state = proj.readState(name);
  if (!state) return res.status(404).json({ error: "프로젝트를 찾을 수 없습니다" });

  const summaries = pipeline.fileSummaries(name);
  res.json({
    state,
    csvExists: fs.existsSync(proj.csvDir(name)),
    totalBlocks: summaries.reduce((a, s) => a + s.blockCount, 0),
    translatedBlocks: summaries.reduce((a, s) => a + s.translatedCount, 0),
    buildExists: fs.existsSync(proj.buildDir(name)),
    outputExists: fs.existsSync(proj.outputDir(name)) && fs.readdirSync(proj.outputDir(name)).some((f) => f.endsWith(".nds")),
  });
});

module.exports = router;

File: webtool/server_py/routes/project.py
"""Python port of webtool/server/routes/project.js - see that file for the
original route contract this mirrors byte-for-byte (paths, JSON shapes)."""
import os
from datetime import datetime, timezone

from flask import Blueprint, jsonify, request

import pipeline
import project as proj

bp = Blueprint("project", __name__)


def _now_iso():
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


@bp.post("/unpack")
def unpack():
    project_name = request.form.get("projectName")
    file = request.files.get("romFile")
    if not file or not project_name:
        return jsonify({"error": "romFile, projectName은 필수입니다"}), 400
    try:
        proj.assert_valid_name(project_name)
    except ValueError as ex:
        return jsonify({"error": str(ex)}), 400
    if not file.filename.lower().endswith(".nds"):
        return jsonify({"error": f".nds 롬 파일이 아닙니다: {file.filename}"}), 400

    try:
        rom_path = proj.original_rom_path(project_name)
        os.makedirs(proj.project_dir(project_name), exist_ok=True)
        file.save(rom_path)

        unpack_dir = proj.unpack_dir(project_name)
        os.makedirs(unpack_dir, exist_ok=True)
        proj.run_nitro_packer(["unpack", "-r", rom_path, "-o", unpack_dir, "-n", project_name])

        script_dir = proj.script_dir(project_name)
        mes_count = 0
        if os.path.exists(script_dir):
            mes_count = sum(1 for f in os.listdir(script_dir) if f.lower().endswith(".mes"))

        state = proj.write_state(project_name, {
            "romFileName": file.filename,
            "romPath": rom_path,
            "createdAt": _now_iso(),
        })

        return jsonify({"ok": True, "state": state, "mesCount": mes_count})
    except Exception as ex:
        return jsonify({"error": str(ex)}), 500


@bp.get("/status")
def status():
    name = request.args.get("name")
    if not name:
        return jsonify({"error": "name은 필수입니다"}), 400
    state = proj.read_state(name)
    if not state:
        return jsonify({"error": "프로젝트를 찾을 수 없습니다"}), 404

    summaries = pipeline.file_summaries(name)
    output_dir = proj.output_dir(name)
    output_exists = os.path.exists(output_dir) and any(
        f.endswith(".nds") for f in os.listdir(output_dir)
    )
    return jsonify({
        "state": state,
        "csvExists": os.path.exists(proj.csv_dir(name)),
        "totalBlocks": sum(s["blockCount"] for s in summaries),
        "translatedBlocks": sum(s["translatedCount"] for s in summaries),
        "buildExists": os.path.exists(proj.build_dir(name)),
        "outputExists": output_exists,
    })
